ledger-service/utils/src/main/scala/com/digitalasset/ledger/service/LedgerReader.scala

```diff
@@ -22,23 +22,44 @@ import Scalaz._
 
 object LedgerReader {
 
+  // TODO(Leo): need something stronger
+  type Error = String
+
+  // PackageId -> Interface
   type PackageStore = Map[String, Interface]
 
+  val UpToDate: Future[Error \/ Option[PackageStore]] =
+    Future.successful(\/-(None))
+
   import scala.concurrent.ExecutionContext.Implicits.global
 
-  def createPackageStore(packageClient: PackageClient): Future[String \/ PackageStore] = {
+  def createPackageStore(packageClient: PackageClient): Future[Error \/ PackageStore] =
+    reloadPackageStore(packageClient)(Set.empty).map(x => x.map(_.getOrElse(Map.empty)))
+
+  /**
+    * @return [[UpToDate]] if packages did not change
+    */
+  def reloadPackageStore(client: PackageClient)(
+      prevPackageIds: Set[String]): Future[Error \/ Option[PackageStore]] = {
     for {
-      packageIds <- packageClient.listPackages().map(_.packageIds)
-      packageResponses <- Future
-        .sequence(packageIds.map(packageClient.getPackage))
-        .map(_.toList)
-    } yield {
-      createPackageStoreFromArchives(packageResponses)
-    }
+      newPackageIds <- client.listPackages().map(_.packageIds.toList)
+      result <- loadIfChanged(client)(prevPackageIds, newPackageIds)
+    } yield result
+  }
+
+  private def loadIfChanged(client: PackageClient)(
+      prevIds: Set[String],
+      newIds: List[String]): Future[Error \/ Option[PackageStore]] = {
+    if (prevIds =/= newIds.toSet)
+      newIds
+        .traverse(id => client.getPackage(id))
```

Comment:
This looks like you are fetching all packages if the ids changed instead of only the ones that are new.

Reply:
Good point. I think @leo-da, it is fine to assume packages are never unloaded from the ledger.

Reply:
yeah, if it never gets removed from the ledger, loading a diff should work

Reply:
chatted with @remyhaemmerle-da we agreed that it is fine to reload the entire thing. We currently don't have `InterfaceReader` logic that allows loading diffs.

Reply:
Let’s say I have a ledger that has several 50 megabytes DARs (not unlikely given our current DAR sizes and upgrades). You definitely don’t want to fetch this every time someone uploads a new dar.

Reply:
sure... but this can and will be done in a separate PR. More refactoring needed.

Reply:
happened to be easier that it looked initially, so loading only new package IDs.